vcd/resource_vcd_dnat.go

```diff
@@ -167,27 +168,56 @@ func resourceVcdDNATCreate(d *schema.ResourceData, meta interface{}) error {
 func resourceVcdDNATRead(d *schema.ResourceData, meta interface{}) error {
 	vcdClient := meta.(*VCDClient)
 
-	e, err := vcdClient.GetEdgeGatewayFromResource(d, "edge_gateway")
+	edgeGateway, err := vcdClient.GetEdgeGatewayFromResource(d, "edge_gateway")
 
 	if err != nil {
 		return fmt.Errorf(errorUnableToFindEdgeGateway, err)
 	}
 
+	// Terraform refresh won't work if Rule was edit in advanced edge gateway UI. vCD API uses <tag> elements to map edge gtw IDs
+	// and UI will reset the tag element on update.
+
 	var found bool
 
-	providedNetworkName := d.Get("network_name")
-	if nil != providedNetworkName && providedNetworkName.(string) != "" {
-		for _, r := range e.EdgeGateway.Configuration.EdgeGatewayServiceConfiguration.NatService.NatRule {
-			if r.RuleType == "DNAT" && r.GatewayNatRule.Interface.Name == d.Get("network_name") &&
-				r.GatewayNatRule.OriginalIP == d.Get("external_ip").(string) &&
-				r.GatewayNatRule.OriginalPort == getPortString(d.Get("port").(int)) {
-				found = true
-				d.Set("internal_ip", r.GatewayNatRule.TranslatedIP)
-			}
+	networkName := d.Get("network_name")
+	if nil != networkName && networkName.(string) != "" {
+		natRule, err := edgeGateway.GetNatRule(d.Id())
+		if err != nil {
+			log.Printf("rule %s (stored in <tag> in Advanced GW case) not found: %s. Removing from state.", d.Id(), err)
+			d.SetId("")
+		}
+
+		portInt, _ := strconv.Atoi(natRule.GatewayNatRule.OriginalPort)
+		translatedPortInt, _ := strconv.Atoi(natRule.GatewayNatRule.TranslatedPort)
+
+		d.Set("description", natRule.Description)
+		d.Set("external_ip", natRule.GatewayNatRule.OriginalIP)
+		d.Set("port", portInt)
+		d.Set("internal_ip", natRule.GatewayNatRule.TranslatedIP)
+		d.Set("translated_port", translatedPortInt)
+		d.Set("protocol", natRule.GatewayNatRule.Protocol)
+		d.Set("icmp_sub_type", natRule.GatewayNatRule.IcmpSubType)
+		d.Set("network_name", natRule.GatewayNatRule.Interface.Name)
+
+		orgVdcNetwork, _ := getOrgVdcNetwork(d, vcdClient, natRule.GatewayNatRule.Interface.Name)
+		if orgVdcNetwork != nil {
+			d.Set("network_type", "org")
+			found = true
+		}
+		externalNetwork, extNetwErr := govcd.GetExternalNetwork(vcdClient.VCDClient, natRule.GatewayNatRule.Interface.Name)
+		if extNetwErr != nil && strings.Contains(extNetwErr.Error(), "could not find external network named") {
+			d.Set("network_type", "ext")
+			found = true
+		}
+		if orgVdcNetwork != nil && extNetwErr == nil {
```

Comment:
A code readability suggestion:

`if orgVdcNetwork != nil && extNetwErr == nil {`
->
`if orgVdcNetwork != nil && externalNetwork != nil {`

Reply:
I can't if you check what returns GetExternalNet - it's a mess. So after spending a lot time, I found easiest way to match what  I need is err.

Reply:
I haven't realized that. Do I get correctly, that `GetExternalNetwork` may return some string that is **not** extnerla network in case of error?

Reply:
It can return empty ExternalNetwork{} or govcd.ExternalNetwork with empty types.ExternalNetwork or something else or error or error with message not found

Reply:
OK, marking this conversation as resolved. CC @dataclouder to have this in mind in the context of `Get` functions "harmonization".